lib/server.js

```diff
@@ -171,23 +194,39 @@ Server.prototype._start = function (config, launcher, preprocess, fileList,
       self._injector.invoke(watcher.watch)
     }
 
-    webServer.listen(config.port, config.listenAddress, function () {
-      self.log.info('Karma v%s server started at %s//%s:%s%s', constant.VERSION,
-        config.protocol, config.listenAddress, config.port, config.urlRoot)
+    var startWebServer = function () {
+      webServer.listen(config.port, config.listenAddress, function () {
+        self.log.info('Karma v%s server started at %s//%s:%s%s', constant.VERSION,
+          config.protocol, config.listenAddress, config.port, config.urlRoot)
 
-      self.emit('listening', config.port)
-      if (config.browsers && config.browsers.length) {
-        self._injector.invoke(launcher.launch, launcher).forEach(function (browserLauncher) {
-          singleRunDoneBrowsers[browserLauncher.id] = false
-        })
-      }
-      var noLoadErrors = self.loadErrors.length
-      if (noLoadErrors > 0) {
-        self.log.error('Found %d load error%s', noLoadErrors, noLoadErrors === 1 ? '' : 's')
-        process.exitCode = 1
-        process.kill(process.pid, 'SIGINT')
-      }
-    })
+        self.emit('listening', config.port)
+        if (config.browsers && config.browsers.length) {
+          self._injector.invoke(launcher.launch, launcher).forEach(function (browserLauncher) {
+            singleRunDoneBrowsers[browserLauncher.id] = false
+          })
+        }
+        var noLoadErrors = self.loadErrors.length
+        if (noLoadErrors > 0) {
+          self.log.error('Found %d load error%s', noLoadErrors, noLoadErrors === 1 ? '' : 's')
+          process.exitCode = 1
+          process.kill(process.pid, 'SIGINT')
+        }
+      })
+    }
+
+    // Check if the static files haven't been compiled
+    if (!(fs.existsSync(karmaJsPath) && fs.existsSync(contextJsPath))) {
+      self.log.info('Front-end scripts not present. Compiling...')
+      var mainPromise = bundleResource(path.join(__dirname, '/../client/main.js'), karmaJsPath)
+      var contextPromise = bundleResource(path.join(__dirname, '/../context/main.js'), contextJsPath)
+      Promise.all([mainPromise, contextPromise]).then(() => {
+        startWebServer()
+      }, (error) => {
+        self.log.error(error)
```

Comment:
- please use catch to handle the errors
- this needs to abort on error because if this fails it means we are unable to start properly

Reply:
Should it just throw the error? Not sure what the correct way to trigger program exit is here.

Reply:
I think exiting like it is done here: https://github.com/EzraBrooks/karma/blob/ba3da6824371c8824c6b5976b55048e54f0a9924/lib/server.js#L210-L212 should be the right course of action